File: src/com/upokecenter/cbor/Base64.java
package com.upokecenter.cbor;
/*
Written by Peter O. in 2014.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */

import com.upokecenter.util.*;

    /**
     * Description of Base64.
     */
  final class Base64 {
private Base64() {
}
    private static final String Base64URL = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
    private static final String Base64Classic = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    /**
     * Not documented yet.
     * @param str A StringBuilder object.
     * @param data A byte array.
     * @param padding A Boolean object.
     */
    public static void ToBase64(StringBuilder str, byte[] data, boolean padding) {
      if (data == null) {
        throw new NullPointerException("data");
      }
      ToBase64(str, data, 0, data.length, Base64Classic, padding);
    }

    public static void ToBase64URL(StringBuilder str, byte[] data, boolean padding) {
      if (data == null) {
        throw new NullPointerException("data");
      }
      ToBase64(str, data, 0, data.length, Base64URL, padding);
    }

    public static void ToBase64(StringBuilder str, byte[] data, int offset, int count, boolean padding) {
      ToBase64(str, data, offset, count, Base64Classic, padding);
    }

    public static void ToBase64URL(StringBuilder str, byte[] data, int offset, int count, boolean padding) {
      ToBase64(str, data, offset, count, Base64URL, padding);
    }

    public static String ToBase64String(byte[] data, boolean padding) {
      if (data == null) {
        throw new NullPointerException("data");
      }
      return ToBase64String(data, 0, data.length, padding);
    }

    public static String ToBase64URLString(byte[] data, boolean padding) {
      if (data == null) {
        throw new NullPointerException("data");
      }
      return ToBase64String(data, 0, data.length, padding);
    }

    public static String ToBase64String(byte[] data, int offset, int count, boolean padding) {
      StringBuilder builder = new StringBuilder();
      ToBase64(builder, data, offset, count, Base64Classic, padding);
      return builder.toString();
    }

    public static String ToBase64URLString(byte[] data, int offset, int count, boolean padding) {
      StringBuilder builder = new StringBuilder();
      ToBase64(builder, data, offset, count, Base64Classic, padding);
      return builder.toString();
    }

    private static void ToBase64(StringBuilder str, byte[] data, int offset, int count, String alphabet, boolean padding) {
      if (data == null) {
        throw new NullPointerException("data");
      }
      if (offset < 0) {
        throw new IllegalArgumentException("offset (" + Integer.toString((int)offset) + ") is less than " + "0");
      }
      if (offset > data.length) {
        throw new IllegalArgumentException("offset (" + Integer.toString((int)offset) + ") is more than " + Integer.toString((int)data.length));
      }
      if (count < 0) {
        throw new IllegalArgumentException("count (" + Integer.toString((int)count) + ") is less than " + "0");
      }
      if (count > data.length) {
        throw new IllegalArgumentException("count (" + Integer.toString((int)count) + ") is more than " + Integer.toString((int)data.length));
      }
      if (data.length - offset < count) {
        throw new IllegalArgumentException("data's length minus " + offset + " (" + Integer.toString((int)(data.length - offset)) + ") is less than " + Integer.toString((int)count));
      }
      int length = offset + count;
      int i = offset;
      for (i = offset; i < (length - 2); i += 3) {
        str.append(alphabet.charAt((data[i] >> 2) & 63));
        str.append(alphabet.charAt(((data[i] & 3) << 4) + ((data[i + 1] >> 4) & 15)));
        str.append(alphabet.charAt(((data[i + 1] & 15) << 2) + ((data[i + 2] >> 6) & 3)));
        str.append(alphabet.charAt(data[i + 2] & 63));
      }
      int lenmod3 = count % 3;
      if (lenmod3 != 0) {
        i = length - lenmod3;
        str.append(alphabet.charAt((data[i] >> 2) & 63));
        if (lenmod3 == 2) {
          str.append(alphabet.charAt(((data[i] & 3) << 4) + ((data[i + 1] >> 4) & 15)));
          str.append(alphabet.charAt((data[i + 1] & 15) << 2));
          if (padding) {
            str.append("=");
          }
        } else {
          str.append(alphabet.charAt((data[i] & 3) << 4));
          if (padding) {
            str.append("==");
          }
        }
      }
    }
  }

File: Cbor/Base64.cs
/*
Written by Peter O. in 2014.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://upokecenter.com/d/
 */
using System;
using System.IO;
using System.Text;

namespace PeterO.Cbor {
  internal static class Base64
  {
    private const string Base64URL =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_ ";

    private const string Base64Classic =
      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/ ";

    public static void ToBase64(StringBuilder str, byte[] data, bool padding) {
      if (data == null) {
        throw new ArgumentNullException("data");
      }
      ToBase64(str, data, 0, data.Length, Base64Classic, padding);
    }

    public static void ToBase64URL(
      StringBuilder str,
      byte[] data,
      bool padding) {
      if (data == null) {
        throw new ArgumentNullException("data");
      }
      ToBase64(str, data, 0, data.Length, Base64URL, padding);
    }

    public static void ToBase64(
      StringBuilder str,
      byte[] data,
      int offset,
      int count,
      bool padding) {
      ToBase64(str, data, offset, count, Base64Classic, padding);
    }

    public static void ToBase64URL(
      StringBuilder str,
      byte[] data,
      int offset,
      int count,
      bool padding) {
      ToBase64(str, data, offset, count, Base64URL, padding);
    }

    public static void WriteBase64(
      Stream outputStream,
      byte[] data,
      int offset,
      int count,
      bool padding) {
      WriteBase64(outputStream, data, offset, count, Base64Classic, padding);
    }

    public static void WriteBase64URL(
      Stream outputStream,
      byte[] data,
      int offset,
      int count,
      bool padding) {
      WriteBase64(outputStream, data, offset, count, Base64URL, padding);
    }

    public static string ToBase64String(byte[] data, bool padding) {
      if (data == null) {
        throw new ArgumentNullException("data");
      }
      return ToBase64String(data, 0, data.Length, padding);
    }

    public static string ToBase64URLString(byte[] data, bool padding) {
      if (data == null) {
        throw new ArgumentNullException("data");
      }
      return ToBase64String(data, 0, data.Length, padding);
    }

    public static string ToBase64String(
      byte[] data,
      int offset,
      int count,
      bool padding) {
      var builder = new StringBuilder();
      ToBase64(builder, data, offset, count, Base64Classic, padding);
      return builder.ToString();
    }

    public static string ToBase64URLString(
      byte[] data,
      int offset,
      int count,
      bool padding) {
      var builder = new StringBuilder();
      ToBase64(builder, data, offset, count, Base64Classic, padding);
      return builder.ToString();
    }

    private static void ToBase64(
      StringBuilder str,
      byte[] data,
      int offset,
      int count,
      string alphabet,
      bool padding) {
      if (str == null) {
        throw new ArgumentNullException("str");
      }
      if (data == null) {
        throw new ArgumentNullException("data");
      }
      if (offset < 0) {
        throw new ArgumentException(
          "offset (" + offset + ") is less than " + "0 ");
      }
      if (offset > data.Length) {
        throw new ArgumentException(
          "offset (" + offset + ") is more than " + data.Length);
      }
      if (count < 0) {
     throw new ArgumentException("count (" + count + ") is less than " +
          "0 ");
      }
      if (count > data.Length) {
        throw new ArgumentException(
          "count (" + count + ") is more than " + data.Length);
      }
      if (data.Length - offset < count) {
        throw new ArgumentException("data's length minus " + offset + " (" +
          (data.Length - offset) +
                                    ") is less than " + count);
      }
      int length = offset + count;
      int i = offset;
      for (i = offset; i < (length - 2); i += 3) {
        str.Append(alphabet[(data[i] >> 2) & 63]);
        str.Append(alphabet[((data[i] & 3) << 4) + ((data[i + 1] >> 4) & 15)]);
        str.Append(alphabet[((data[i + 1] & 15) << 2) + ((data[i + 2] >> 6) &
                                                         3)]);
        str.Append(alphabet[data[i + 2] & 63]);
      }
      int lenmod3 = count % 3;
      if (lenmod3 != 0) {
        i = length - lenmod3;
        str.Append(alphabet[(data[i] >> 2) & 63]);
        if (lenmod3 == 2) {
          str.Append(alphabet[((data[i] & 3) << 4) + ((data[i + 1] >> 4) &
                                                      15)]);
          str.Append(alphabet[(data[i + 1] & 15) << 2]);
          if (padding) {
            str.Append("=");
          }
        } else {
          str.Append(alphabet[(data[i] & 3) << 4]);
          if (padding) {
            str.Append("==");
          }
        }
      }
    }

    private static void WriteBase64(
      Stream outputStream,
      byte[] data,
      int offset,
      int count,
      string alphabet,
      bool padding) {
      if (outputStream == null) {
        throw new ArgumentNullException("outputStream");
      }
      if (offset < 0) {
   throw new ArgumentException("offset (" + offset + ") is less than " +
          "0 ");
      }
      if (offset > data.Length) {
        throw new ArgumentException("offset (" + offset + ") is more than " +
          data.Length);
      }
      if (count < 0) {
     throw new ArgumentException("count (" + count + ") is less than " +
          "0 ");
      }
      if (count > data.Length) {
        throw new ArgumentException("count (" + count + ") is more than " +
          data.Length);
      }
      if (data.Length - offset < count) {
        throw new ArgumentException("data's length minus " + offset + " (" +
          (data.Length - offset) +
                                    ") is less than " + count);
      }
      int length = offset + count;
      int i = offset;
      var buffer = new byte[4];
      for (i = offset; i < (length - 2); i += 3) {
        buffer[0] = (byte)alphabet[(data[i] >> 2) & 63];
        buffer[1] = (byte)alphabet[((data[i] & 3) << 4) +
                                   ((data[i + 1] >> 4) &
                                    15)];
        buffer[2] = (byte)alphabet[((data[i + 1] & 15) << 2) + ((data[i +
                                                                      2] >> 6) &
                                                                3)];
        buffer[3] = (byte)alphabet[data[i + 2] & 63];
        outputStream.Write(buffer, 0, 4);
      }
      int lenmod3 = count % 3;
      if (lenmod3 != 0) {
        i = length - lenmod3;
        buffer[0] = (byte)alphabet[(data[i] >> 2) & 63];
        if (lenmod3 == 2) {
          buffer[1] = (byte)alphabet[((data[i] & 3) << 4) + ((data[i + 1] >>
                                                              4) &
                                                             15)];
          buffer[2] = (byte)alphabet[(data[i + 1] & 15) << 2];
          if (padding) {
            buffer[3] = (byte)'=';
            outputStream.Write(buffer, 0, 4);
          } else {
            outputStream.Write(buffer, 0, 3);
          }
        } else {
          buffer[1] = (byte)alphabet[(data[i] & 3) << 4];
          if (padding) {
            buffer[2] = (byte)'=';
            buffer[3] = (byte)'=';
            outputStream.Write(buffer, 0, 4);
          } else {
            outputStream.Write(buffer, 0, 2);
          }
        }
      }
    }
  }
}
